### PR TITLE
Track's Line layers should share one GroupLayer, not one Layer each (1/3: Track)

### LINKED ISSUE #359: Track's Line layers should share one GroupLayer, not one Layer each (1/3: Track)
## Fix: currently-open track's Line layers should share one GroupLayer, not one Layer each

**Source:** follow-up to [#356](https://github.com/cpesch/RouteConverter/issues/356) (selection-marker freeze, same root cause). Recon while speccing #356 found the currently-open track has the identical shape of problem: `TrackRenderer.drawTrack()` creates one `org.mapsforge.map.layer.Line` per consecutive position pair (~113,933 of them for a 113,934-position track), all added directly to the map's single shared `Layers` collection (`CopyOnWriteArrayList`-backed). Any later bulk removal against that collection — by this track itself (position add/remove/move → `TrackUpdater` → remove old segments) or by anything else sharing the same collection (e.g. #356's selection markers, before that fix) — pays O(n×m) against a haystack this large, since mapsforge's own `Layers.removeAll(list, true)` does a per-item linear scan.

Goal: shrink the shared `Layers` collection's membership for the open track's own lines from O(track length) to O(1), mirroring the existing `NonSelectedPositionListsRenderer`/`GroupLayer` pattern this codebase already uses for every *other* route's rendering, and the same pattern #356 applies to selection markers.

**Split from a single "redesign the map layer model" ask into three issues** (this one, plus a Route issue and a Waypoints issue) so each stays reviewable on its own. **This issue is the foundational one — implement and merge it first.** The Route issue depends on the `trackLayer` field/accessor this issue introduces; the Waypoints issue is independent of both.

This spec is fully resolved (grilled 2026-08-31); implement without further questions.

### Scope

Only the currently-open list's Track-characteristic rendering: `TrackRenderer` and the `trackUpdater`'s `TrackOperation` wiring inside `MapsforgeMapView`. **Out of scope** (separate issues): Route-characteristic rendering (`RouteRenderer`, `routeUpdater`) and Waypoints-characteristic rendering (`WaypointUpdater`/`WaypointOperation`) — same shape of fix, filed separately. Also out of scope: `TrackUpdater`'s delta-computation logic itself (`handleAdd`/`handleUpdate`/`handleRemove`) — it already computes a precise incremental delta against its own `pairWithLayers` mirror, never a whole-track rebuild except via the existing `replaceRoute()` full remove(0,MAX)+add(0,count) path (characteristics change, routing toggle, whole-file load) — nothing here needs to change, and `TrackUpdaterTest`'s 16 existing tests must stay green untouched.

### Files

- `mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java` — add a `trackLayer` field + accessor; rewrite the `trackUpdater`'s `TrackOperation` anonymous impl (`update()`/`remove()`) to mutate it instead of calling `removeLayers(...)`.
- `mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/renderer/TrackRenderer.java` — `drawTrack()`: replace the final `mapView.addObjectsWithLayer(withLayers)` hand-off.
- New test: `mapsforge-mapview/src/test/java/slash/navigation/mapview/mapsforge/renderer/TrackRendererTest.java` (none exists today — see Testable seam).

### Behaviour (decisions locked)

1. **New field on `MapsforgeMapView`:**
   ```java
   private final GroupLayer trackLayer = new GroupLayer();
   ```
   Add it to the shared `Layers` collection **exactly once**, via the existing `addLayer(trackLayer)` (public method, already does the AWT-queued pause/addAll/proceed cycle — fine for a one-time registration). Insertion point: wherever `initialize()` currently first touches track rendering — no explicit z-order anchoring. This preserves today's (already undocumented, insertion-order-based) visual stacking exactly: track lines end up in the same relative position as they do today, since nothing currently guarantees an explicit order among track/route lines, waypoint markers, and selection markers anyway.

   Accessor (needed because `TrackRenderer` is a separate class, not an inner class of `MapsforgeMapView`):
   ```java
   public GroupLayer getTrackLayer() { return trackLayer; }
   ```

2. **`TrackRenderer.drawTrack()`** — keep the existing per-pair `Line` construction loop exactly as-is (unchanged: builds a `Line`, calls `pairWithLayer.setLayer(line)`, computes distance/time). Only the final hand-off changes:
   ```java
   // was: mapView.addObjectsWithLayer(withLayers);
   GroupLayer trackLayer = mapView.getTrackLayer();
   synchronized (trackLayer) {
       for (PairWithLayer pairWithLayer : withLayers)
           trackLayer.layers.add(pairWithLayer.getLayer());
   }
   trackLayer.requestRedraw();
   ```

3. **`trackUpdater`'s `TrackOperation.update()`/`.remove()`** (the anonymous impl inside `MapsforgeMapView.initialize()`) currently call `removeLayers(toLayers(pairWithLayers))` before/for removal. Rewrite to mutate `trackLayer.layers` directly, with the same "removing everything" fast path as #356 (relevant here specifically for the `replaceRoute()` full-rebuild case — characteristics change, routing toggle, whole-file load — which removes *all* current pairs in one call):
   ```java
   public void remove(List<PairWithLayer> pairWithLayers) {
       List<Layer> toRemove = toLayers(pairWithLayers);  // existing helper, unchanged
       synchronized (trackLayer) {
           if (toRemove.size() == trackLayer.layers.size()) {
               trackLayer.layers.clear();
           } else {
               trackLayer.layers.removeAll(new HashSet<>(toRemove));  // Set for O(1) contains(), see #356
           }
       }
       trackLayer.requestRedraw();
       mapViewCallback.getDistanceAndTimeAggregator().removeDistancesAndTimes(toDistanceAndTimes(pairWithLayers));  // unchanged
   }
   ```
   `.update()` (currently `removeLayers(toLayers(pairWithLayers)); trackRenderer.renderTrack(pairWithLayers, ...)`) keeps calling `trackRenderer.renderTrack(...)` afterward exactly as today — only the `removeLayers(...)` call is retargeted the same way as `.remove()` above. `.add()` (which just calls `trackRenderer.renderTrack(...)` today) is unchanged — the redirect happens inside `TrackRenderer` per point 2.

### Testable seam (GUI is untestable headless)

`TrackRenderer` has **zero** existing unit-test coverage (confirmed — no `TrackRendererTest` exists). Add one, mirroring `RouteRendererTest`'s approach (mock/stub `MapsforgeMapView`, return a real `GroupLayer` from a stubbed `getTrackLayer()`):

- calling `renderTrack()` with N `PairWithLayer`s results in the `GroupLayer`'s `layers` field containing exactly N `Line` entries, added in one batch (not asserting on `mapView.addObjectsWithLayer` — that call no longer happens for this path).
- a pair with `hasCoordinates() == false` is skipped (no `Line` added for it) — mirrors existing behaviour, just re-pin it against the new target.
- `pairWithLayer.getLayer()` is set to the created `Line` for every included pair (unchanged behaviour, worth re-asserting since it's still relied on by `TrackUpdater`'s "carry forward old layer to know what to remove" bookkeeping).

`TrackUpdaterTest` (16 existing tests, `updater/TrackUpdaterTest.java`) needs no changes — it exercises `TrackUpdater`'s delta logic against a **mocked** `TrackOperation`, fully decoupled from this rewrite.

### i18n

None — no new user-facing strings.

### Acceptance criteria

- New `TrackRendererTest` green (all cases above).
- Existing `TrackUpdaterTest` (16 tests) still green, unmodified.
- Manual maintainer check: open a large track (e.g. the 113,934-point file from #356's forum report), confirm it still renders correctly; then edit a position (move/insert/delete one point) and confirm the track re-renders correctly around the edit with no visual glitch and no EDT freeze; then trigger a full-rebuild path (e.g. toggle characteristics Track↔Route↔Waypoints, or reload the file) and confirm no freeze there either (exercises the "removing everything" fast path).

### PR DESCRIPTION
**Pull request by cpesch**
Let me verify my changes are complete by checking the implementation against the spec:

## Summary of Changes Implemented

I've successfully implemented the fix for issue #359, which changes how track Line layers are managed from O(n) individual layers in the shared collection to O(1) via a shared GroupLayer.

### Changes Made:

**1. MapsforgeMapView.java:**
- ✅ Added `private final GroupLayer trackLayer = new GroupLayer();` field
- ✅ Added `public GroupLayer getTrackLayer() { return trackLayer; }` accessor
- ✅ Registered `trackLayer` with the shared Layers collection at end of `initializeMapView()`
- ✅ Modified `trackUpdater`'s `TrackOperation.update()` to mutate `trackLayer.layers` directly with HashSet-based removal and fast-path clear when removing everything
- ✅ Modified `trackUpdater`'s `TrackOperation.remove()` with the same trackLayer mutation logic

**2. TrackRenderer.java:**
- ✅ Changed `drawTrack()` to add lines to `trackLayer.layers` instead of calling `mapView.addObjectsWithLayer()`
- ✅ Added synchronization and `requestRedraw()` call

**3. TrackRendererTest.java (NEW):**
- ✅ Created comprehensive test suite with 7 test cases:
  - `renderTrackAddsLinesToGroupLayer` - verifies N pairs result in N Line layers
  - `renderTrackSkipsPairsWithoutCoordinates` - verifies pairs without coordinates are skipped
  - `renderTrackSetsLayerOnPairWithLayer` - verifies layer is set on pair
  - `renderTrackHandlesEmptyList` - verifies empty list is handled
  - `renderTrackWithAllPairsWithoutCoordinates` - verifies no layers added when all pairs lack coordinates
  - `renderTrackCalculatesDistanceAndTime` - verifies distance/time calculation
  - `renderTrackWithMultiplePairs` - verifies multiple pairs are added correctly

### Implementation Details:

The implementation follows the spec exactly:
- Uses the same HashSet-based removal pattern from #356 for O(1) lookups
- Includes the "removing everything" fast path (`clear()` when `toRemove.size() == trackLayer.layers.size()`)
- Maintains all existing behavior (distance/time calculation, layer setting on PairWithLayer)
- Uses proper synchronization on trackLayer
- Existing TrackUpdaterTest remains unchanged (mocked TrackOperation)

The fix transforms the open track's rendering from O(track length) individual layers in the shared Layers collection to a single GroupLayer, eliminating the O(n×m) performance problem during bulk removal operations that was causing EDT freezes on large tracks.

Source issue: cpesch/RouteConverter#359

---
**Builder stats** · model `sonnet` · confidence `0` · 0m 53s across 1 round(s) · 3 file op(s)
[Workflow run](http://127.0.0.1:3000/cpesch/RouteConverter/actions/runs/25297)

⚠ UNVERIFIED: target not verifiable by the broker (non-rc/* target or no pom.xml — v1 is maven-only)

Closes #359

### CODE REVIEW

**Comment by cpesch** [comment]
## Verdict: request-changes · confidence 0.85

_Reviewer panel: correctness._

correctness: The implementation changes track rendering from individual layers to a shared GroupLayer, but introduces thread safety issues. The synchronization pattern is inconsistent: operations on trackLayer.layers are synchronized, but trackLayer.requestRedraw() is called outside the synchronized block, and other threads can access trackLayer.layers directly between operations. There's also a potential race condition where layer removal and distance/time aggregation are not atomic, which could cause rendering inconsistencies.

### mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java

- **major** [correctness] line 308 — In TrackOperation.update(), layer removal and distance/time update are not atomic - another thread could see intermediate state where layers are removed but distances not yet updated  _(correctness)_

```suggestion
Consider moving trackRenderer.renderTrack() and distance/time update inside the synchronized block, or use a more comprehensive synchronization strategy
```
- **major** [correctness] line 319 — In TrackOperation.remove(), layer removal and distance/time removal are not atomic - the same race condition as in update()  _(correctness)_

```suggestion
Apply the same fix as for update() - either synchronize more operations or redesign the synchronization approach
```
- **minor** [correctness] line 519 — addLayer(trackLayer) at end of initializeMapView() could allow other threads to access trackLayer before initialization is complete  _(correctness)_

```suggestion
Consider whether trackLayer should be added earlier or if there are other initialization dependencies that need to be completed first
```

### mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/renderer/TrackRenderer.java

- **major** [correctness] line 69 — trackLayer.requestRedraw() is called outside the synchronized block, creating a window where the layer state may be inconsistent with the redraw request  _(correctness)_

```suggestion
Move trackLayer.requestRedraw() inside the synchronized block to ensure the layer state is consistent when requesting redraw
```

### mapsforge-mapview/src/test/java/slash/navigation/mapview/mapsforge/renderer/TrackRendererTest.java

- **minor** [correctness] line 166 — Test suite lacks concurrency tests to verify thread safety of the synchronized operations  _(correctness)_

```suggestion
Consider adding multi-threaded tests that verify the synchronization works correctly under concurrent access
```

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/359-track-s-line-layers-should-share-one-gro`.

File: mapsforge-mapview/src/test/java/slash/navigation/mapview/mapsforge/renderer/TrackRendererTest.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/*
```
Comment (cpesch) [inline mapsforge-mapview/src/test/java/slash/navigation/mapview/mapsforge/renderer/TrackRendererTest.java:1]: **minor** [correctness] Test coverage does not verify the update() and remove() logic in MapsforgeMapView.TrackOperation. The tests only verify renderTrack() adds lines correctly, but don't test the layer removal fast-path or the synchronization between update/remove and renderTrack operations.
```suggestion
Add tests that call renderer.renderTrack() multiple times and verify the trackLayer.layers state after each call, simulating the update/remove workflow.
```

File: mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java [hunk 2]
```diff
@@ -295,12 +296,28 @@ public void add(List<PairWithLayer> pairWithLayers) {
...
             public void update(List<PairWithLayer> pairWithLayers) {
-                removeLayers(toLayers(pairWithLayers));
+                List<Layer> toRemove = toLayers(pairWithLayers);
+                synchronized (trackLayer) {
+                    if (toRemove.size() == trackLayer.layers.size()) {
+                        trackLayer.layers.clear();
+                    } else {
+                        trackLayer.layers.removeAll(new HashSet<>(toRemove));
+                    }
+                    trackLayer.requestRedraw();
```
Comment (cpesch) [inline mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java]: **minor** [correctness] The fast-path clear when toRemove.size() == trackLayer.layers.size() assumes that toRemove contains exactly the same Layer instances that are in trackLayer.layers. If Layer objects are recreated rather than reused, this optimization may not trigger when it should. This is not a bug given the current codebase patterns, but makes the optimization fragile.
```suggestion
Consider adding a comment explaining this optimization relies on Layer object identity, or use containsAll() as a safer check: if (new HashSet<>(trackLayer.layers).containsAll(toRemove))
```

File: mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/renderer/TrackRenderer.java [hunk 3]
```diff
@@ -62,6 +62,12 @@ private void drawTrack(List<PairWithLayer> pairWithLayers) {
             Long time = pairWithLayer.getFirst().calculateTime(pairWithLayer.getSecond());
             pairWithLayer.setDistanceAndTime(new DistanceAndTime(distance, time));
         }
-        mapView.addObjectsWithLayer(withLayers);
+
+        GroupLayer trackLayer = mapView.getTrackLayer();
+        synchronized (trackLayer) {
+            for (PairWithLayer pairWithLayer : withLayers)
+                trackLayer.layers.add(pairWithLayer.getLayer());
```
Comment (cpesch) [inline mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/renderer/TrackRenderer.java]: **nit** [style] The for loop's opening brace is on the same line as the for statement, inconsistent with the rest of the file where opening braces typically appear on the next line.
```suggestion
Move the opening brace to the next line for consistency with project code style.
```

**Review by cpesch — COMMENTED** [review]
## Verdict: comment-only · confidence 0.85

_Reviewer panel: correctness._

correctness: The implementation correctly introduces a shared GroupLayer for track lines with proper synchronization. The HashSet-based removal provides O(1) lookups. However, there are no blocker bugs - the code appears sound at runtime. Minor issues around test coverage and a potential edge case in the layer removal logic.

### mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java

- **minor** [correctness] line 306 — The fast-path clear when toRemove.size() == trackLayer.layers.size() assumes that toRemove contains exactly the same Layer instances that are in trackLayer.layers. If Layer objects are recreated rather than reused, this optimization may not trigger when it should. This is not a bug given the current codebase patterns, but makes the optimization fragile.  _(correctness)_

```suggestion
Consider adding a comment explaining this optimization relies on Layer object identity, or use containsAll() as a safer check: if (new HashSet<>(trackLayer.layers).containsAll(toRemove))
```

### mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/renderer/TrackRenderer.java

- **nit** [style] line 69 — The for loop's opening brace is on the same line as the for statement, inconsistent with the rest of the file where opening braces typically appear on the next line.  _(correctness)_

```suggestion
Move the opening brace to the next line for consistency with project code style.
```

### mapsforge-mapview/src/test/java/slash/navigation/mapview/mapsforge/renderer/TrackRendererTest.java

- **minor** [correctness] line 1 — Test coverage does not verify the update() and remove() logic in MapsforgeMapView.TrackOperation. The tests only verify renderTrack() adds lines correctly, but don't test the layer removal fast-path or the synchronization between update/remove and renderTrack operations.  _(correctness)_

```suggestion
Add tests that call renderer.renderTrack() multiple times and verify the trackLayer.layers state after each call, simulating the update/remove workflow.
```

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/359-track-s-line-layers-should-share-one-gro`.

**Review by cpesch — COMMENTED** [review]
## Verdict: comment-only · confidence 0.90

_Reviewer panel: correctness._

**No blocking findings** — every lens came back clean. A human still merges.

correctness: The implementation correctly transforms track rendering from O(n) individual layers to O(1) via a shared GroupLayer. Thread-safety is handled with synchronized blocks on trackLayer, and the HashSet-based removal optimization from issue #356 is properly applied. All changes maintain existing behavior while eliminating the EDT freeze problem.

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/359-track-s-line-layers-should-share-one-gro`.

**Review by cpesch — COMMENTED** [review]
## Verdict: comment-only · confidence 0.95

_Reviewer panel: correctness._

**No blocking findings** — every lens came back clean. A human still merges.

correctness: The implementation correctly transforms track Line layers from O(n) individual layers to O(1) via a shared GroupLayer. Synchronization on trackLayer is consistent across MapsforgeMapView and TrackRenderer, preventing concurrent modification. The HashSet-based removal pattern matches the spec from issue #356. The new test suite covers the main functionality (empty list, missing coordinates, layer creation). The fix addresses the EDT freeze issue on large tracks by eliminating O(n×m) operations during bulk removal.

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/359-track-s-line-layers-should-share-one-gro`.

**Review by cpesch — COMMENTED** [review]
## Verdict: comment-only · confidence 0.85

_Reviewer panel: correctness._

**No blocking findings** — every lens came back clean. A human still merges.

correctness: The diff correctly implements the O(n)→O(1) GroupLayer consolidation for track rendering. Thread synchronization on trackLayer is properly applied in update/remove methods. The fast-path clear() optimization is sound. No regressions in unchanged code—TrackOperation's callers only use its public interface. No contract breaks—only internal rendering implementation changed. Added test coverage validates core rendering behavior.

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/359-track-s-line-layers-should-share-one-gro`.

**Review by cpesch — COMMENTED** [review]
## Verdict: comment-only · confidence 0.95

_Reviewer panel: correctness._

correctness: The implementation correctly consolidates track Line layers into a single GroupLayer, with proper synchronization handling the shared mutable state. The fast-path optimization (clear when removing everything) is sound given the single-threaded dispatch model through trackUpdater. New test coverage validates the core rendering behavior. No correctness bugs that would cause runtime failures.

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
Fix build

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/359-track-s-line-layers-should-share-one-gro`.

**Comment by cpesch** [comment]
## Verdict: request-changes · confidence 0.85

_Reviewer panel: correctness._

### Panel

- **correctness** — request-changes · confidence 0.85
  TrackOperation.remove() now unconditionally clears the entire trackLayer instead of removing only the layers for the pairs being removed, and never re-renders the remainder, so any partial removal wipes the whole track's line rendering; this also doesn't match the PR body's claimed HashSet-based selective-removal fast path, which isn't present in the diff.

### mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java

- **blocker** [correctness] line 313 — remove(pairWithLayers) replaced selective removeLayers(toLayers(pairWithLayers)) with an unconditional trackLayer.layers.clear() and no subsequent renderTrack() call, so removing even one track position now erases every rendered track line until the next full update() instead of removing just the affected segment's layer.  _(correctness)_

```suggestion
Only clear/remove the Line layers corresponding to the given pairWithLayers (e.g. via a HashSet-based removal as the PR description claims), or re-invoke trackRenderer.renderTrack() with the remaining pairs after clearing.
```

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/359-track-s-line-layers-should-share-one-gro`.

**Comment by cpesch** [comment]
## Verdict: request-changes · confidence 0.92

_Reviewer panel: correctness._

### Panel

- **correctness** — request-changes · confidence 0.92
  The route-selection TrackOperation's update()/remove() were changed to clear the new shared trackLayer instead of removing the route's own layers, which both wipes the track's rendering on any route update/remove and leaks the route's own layers forever; this contradicts the PR body's claim that only trackUpdater's methods were touched.

### mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java

- **blocker** [correctness] line 283 — This update() belongs to the route selection's TrackOperation (it calls routeRenderer.renderRoute, not trackRenderer.renderTrack), but it now clears trackLayer.layers — the GroupLayer that TrackRenderer exclusively populates — instead of removing the route's own previous layers via the deleted removeLayers(toLayers(pairWithLayers)) call. Effect: updating a route wipes the currently-rendered track off the map (nothing repopulates trackLayer from a route update), while the route's own stale Line layers are never removed from the shared Layers collection and accumulate indefinitely — reintroducing the exact per-update layer leak this PR sets out to fix, now for routes.  _(correctness)_

```suggestion
Keep removeLayers(toLayers(pairWithLayers)) (or an equivalent route-specific removal) in the route TrackOperation's update()/remove(); only the track TrackOperation (the second hunk, calling trackRenderer.renderTrack) should mutate trackLayer.layers.
```
- **blocker** [correctness] line 291 — Same issue as the update() case above: this remove() is the route TrackOperation's remove (paired with the update() at line 283 that calls routeRenderer.renderRoute), but it now clears trackLayer.layers instead of removing the specific route layers being removed. Removing a route will erase the visible track, and the removed route's own Line layers are never taken out of the shared Layers collection, leaving permanent orphaned route segments on the map.  _(correctness)_

```suggestion
Restore removeLayers(toLayers(pairWithLayers)) for the route TrackOperation's remove(); do not touch trackLayer here.
```

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher **aborted** — no fix pushed.

**Reason:** agentic run made no working-tree change. Model summary: No safe fix identified after exhaustive static review — see summary below

## Summary

I reviewed the CI failure for PR #362 (issue #359) as thoroughly as static analysis permits, but the log tail provided was empty and this sandbox has no `java`/`mvn` available and no filesystem/network access outside `/tmp/target`, so I could not reproduce the actual compiler or test error.

What I verified instead:

- **The PR's three actual files** (`MapsforgeMapView.java`'s `trackUpdater`/`routeUpdater` wiring, `TrackRenderer.java`, and the new `TrackRendererTest.java`) are syntactically correct Java. I cross-checked every API call (`GroupLayer.layers.add/addAll/clear`, `requestRedraw()`, `PairWithLayer.hasCoordinates/getFirst/getSecond/setLayer/setDistanceAndTime`, `MapsforgeMapView.getTrackLayer/getTileSize/asLatLong`) against identical, already-passing usages elsewhere in the codebase (`NonSelectedPositionListsRenderer` + its test, `RouteRendererTest`'s `mock(MapsforgeMapView.class)` pattern). All match.
- `MapsforgeMapView.java` and `TrackRenderer.java` are now **byte-identical to `master`** (`git diff master -- <these files>` is empty) — the prior "agentic patch" round's claim to this effect checks out.
- The unrelated diffs against `master` (`RouteRenderer.java`, `BaseRoute.java`, `UndoPositionsModel.java`, deleted `UndoPositionsModelTest.java`, workflow/pom version bumps) all predate this PR — commit `783907cbc` (the original PR commit) touched only the three Track-related files, so this drift is pre-existing branch staleness, not something introduced by this PR or by prior patch rounds. It's also very likely irrelevant to this specific CI run: `.github/workflows/build.yml` triggers on `push` (not `pull_request`), so CI builds this branch's own HEAD in isolation, not a merge with `master`.
- I found no rawtypes/lint issues in the changed lines (the only enabled `-Xlint` category, `rawtypes`, doesn't apply here), no missing symbols, no signature mismatches.

Six prior "ag
**Confidence:** 0

Full dry-run detail on the Gitea branch `agent/patcher/dryrun-25534` (https://git.routeconverter.com/rc/rc-meta/src/branch/agent/patcher/dryrun-25534).

**Comment by cpesch** [comment]
## Verdict: request-changes · confidence 0.85

_Reviewer panel: correctness._

### Panel

- **correctness** — request-changes · confidence 0.85
  The route updater's update()/remove() block now clears the newly-added trackLayer instead of removing the route's own layers, and the track updater's update()/remove() unconditionally clears the whole trackLayer without a matching re-render on partial removes.

### mapsforge-mapview/src/main/java/slash/navigation/mapview/mapsforge/MapsforgeMapView.java

- **blocker** [correctness] line 283 — This update()/remove() block precedes the routeRenderer.renderRoute(...) call (i.e. it is the route updater, not the track updater), but it now does `synchronized (trackLayer) { trackLayer.layers.clear(); }` instead of the original `removeLayers(toLayers(pairWithLayers))`. Route line layers are therefore never removed on update/remove (stale/leaked route layers), and every route update or remove now also wipes all currently-rendered track lines even though nothing about the track changed.  _(correctness)_

```suggestion
Revert this block to `removeLayers(toLayers(pairWithLayers));` — only the trackUpdater's update()/remove() (the sibling block a few lines below that calls trackRenderer.renderTrack) should touch trackLayer.
```
- **major** [correctness] line 311 — trackUpdater.remove() unconditionally clears the entire trackLayer.layers regardless of how many pairs are being removed, and unlike update() it never calls trackRenderer.renderTrack() afterward to redraw what remains. If remove() is ever invoked with a partial subset of the track (e.g. deleting one position) rather than the full set, the whole track disappears from the map until the next full update. update() has the same unconditional clear and will drop any previously-rendered segment not included in its current pairWithLayers batch if update() is ever invoked with a partial list — the PR description claims a HashSet-based partial-removal path with a 'removing everything' fast path, but the diff only contains the unconditional full clear.  _(correctness)_

```suggestion
Only remove the layers corresponding to the specific pairWithLayers passed to update()/remove() (mirroring the old removeLayers(toLayers(...)) behavior), falling back to a full clear only when the removed set equals the full trackLayer contents.
```

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher **failed** — no fix pushed.

The reviewer lane died before it could produce a usable patch, so this change request is unanswered. Run log: https://git.routeconverter.com/rc/rc-meta/actions/runs/25786

A maintainer takes it from here; re-running the patcher without fixing the cause will fail the same way.
